### PR TITLE
Add Pulsar XBOARD QS Keyboard

### PR DESCRIPTION
**Pull request by skylupin03**
Add "Pulsar XBOARD QS" Keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/pulsar/xboard_qs/keymaps/default/readme.md
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/keymaps/default/readme.md:1]: This file should be removed.

File: keyboards/pulsar/xboard_qs/keymaps/via/readme.md
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/keymaps/via/readme.md:1]: This file should be removed.

File: keyboards/pulsar/xboard_qs/mcuconf.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#undef STM32_PREDIV_VALUE  
+#define STM32_PREDIV_VALUE                  1
+#undef STM32_PLLMUL_VALUE 
+#define STM32_PLLMUL_VALUE                  4
+#undef STM32_USBSW  
+#define STM32_USBSW                         STM32_USBSW_PCLK
+#undef STM32_ST_USE_TIMER   
+#define STM32_ST_USE_TIMER                  3
+
+//#define STM32_USB_USE_USB1                  TRUE
```
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/mcuconf.h:41]:
```suggestion
```

File: keyboards/pulsar/xboard_qs/xboard_qs.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+#ifdef RGB_MATRIX_ENABLE
+#define ___ NO_LED
+
+bool rgb_matrix_indicators_kb(void)  {
+    return rgb_matrix_indicators_user();
+    return TRUE;
+} 
+
```
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/xboard_qs.c:42]:
```suggestion
```

File: keyboards/pulsar/xboard_qs/xboard_qs.c [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    keyboard_pre_init_user();
+}
+
+void keyboard_post_init_kb(void) {
+    keyboard_post_init_user();
+}
+
+void suspend_wakeup_init_kb(void) { // code will run on keyboard wakeup
+    clear_keyboard();
+}
```
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/xboard_qs.c:33]:
```suggestion
```

File: keyboards/pulsar/xboard_qs/rules.mk [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow (WS2811, WS2812, WS2812B, WS2812C, SK6812, SK6812MINI, SK6805, APA102)
+RGB_MATRIX_ENABLE = yes			#
+
+AUDIO_ENABLE = no           # Audio output
+LTO_ENABLE = yes
+
+DYNAMIC_MACRO_ENABLE = yes
+WPM_ENABLE = yes
+
+ENCODER_ENABLE = yes        # Enable Encoder
```
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/rules.mk:19]:
```suggestion
# This file intentionally left blank
```

File: keyboards/pulsar/xboard_qs/info.json [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        "pid": "0x2400",
+        "device_version": "0.0.4"
+    },
+
+    "matrix_pins": {
+        "rows": ["A0", "A1", "A2", "A3", "A4", "A5", "A6", "A7"],
+        "cols": ["C4", "C5", "B0", "B1", "B2", "B10", "B11", "B12", "B13", "B14", "A15", "C6", "C7", "C10", "C11", "C12"]
+    },
+    "diode_direction": "ROW2COL",
+
```
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/info.json:21]:
```suggestion
    "build": {
        "lto": true
    },
    "features": {
        "bootmagic": true,
        "extrakey": true,
        "mousekey": true,
        "nkro": true,
        "encoder": true,
        "rgb_matrix": true
    },
```

File: keyboards/pulsar/xboard_qs/config.h [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#define WAIT_US_TIMER           GPTD3       //  Encoder 및 RGB ANI가 부드러워짐
+
+// Dynamic EEPROM
```
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/config.h:21]: Many of the config options below should be migrated to info.json

File: keyboards/pulsar/xboard_qs/chconf.h
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/chconf.h:1]: This file should be removed

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/pulsar/xboard_qs/keymaps/default/readme.md
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/keymaps/default/readme.md:1]: ok thank you~

File: keyboards/pulsar/xboard_qs/keymaps/via/readme.md
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/keymaps/via/readme.md:1]: Thank you

File: keyboards/pulsar/xboard_qs/mcuconf.h (hunk 1 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/mcuconf.h:41]: OK~ Thank you!

File: keyboards/pulsar/xboard_qs/xboard_qs.c (hunk 2 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/xboard_qs.c:42]: OK~ Thank you!

File: keyboards/pulsar/xboard_qs/xboard_qs.c (hunk 3 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/xboard_qs.c:33]: Thank you for teaching me

File: keyboards/pulsar/xboard_qs/rules.mk (hunk 4 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/rules.mk:19]: Thank you, but,
How can I enable the "Dynamic_MACRO_ENABLE" function?

Comment (adophoxia) [reply keyboards/pulsar/xboard_qs/rules.mk:19]: Is there a reason you need this feature enabled?

File: keyboards/pulsar/xboard_qs/chconf.h
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/chconf.h:1]: Thank you

File: keyboards/pulsar/xboard_qs/config.h (hunk 6 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/config.h:21]: Thank you.
I'll check and fix it as much as I can

File: keyboards/pulsar/xboard_qs/chconf.h
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/chconf.h:1]: from ./lib/chibios/os/hal/include/hal.h:334,
                 from platforms/chibios/platform_deps.h:18,
                 from quantum/quantum.h:18,
                 from keyboards/pulsar/xboard_qs/xboard_qs.c:17:
./lib/chibios/os/hal/ports/STM32/LLD/SYSTICKv1/hal_st_lld.h:178:2: error: #error "ST requires TIM3 but the timer is already used"
  178 | #error "ST requires TIM3 but the timer is already used"
      |  ^~~~~

Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/chconf.h:1]: The error came out as below.
Has the system changed to use TIM3 as the default Timer?

I don't know. Help me
-----------------------------------
Compiling: lib/chibios/os/hal/ports/STM32/LLD/SPIv2/hal_i2s_lld.c                                   [OK]
Compiling: lib/chibios/os/hal/ports/STM32/LLD/SPIv2/hal_spi_v2_lld.c                                [OK]
Compiling: lib/chibios/os/hal/ports/STM32/LLD/SYSTICKv1/hal_st_lld.c                               lib/chibios/os/hal/ports/STM32/LLD/SYSTICKv1/hal_st_lld.c:79:2: error: #error "TIM3 is not a 32bits timer"
   79 | #error "TIM3 is not a 32bits timer"
      |  ^~~~~
 [ERRORS]

File: keyboards/pulsar/xboard_qs/mcuconf.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+#undef STM32_PLLSRC 
+#define STM32_PLLSRC                        STM32_PLLSRC_HSE
+#undef STM32_PREDIV_VALUE  
+#define STM32_PREDIV_VALUE                  1
+#undef STM32_PLLMUL_VALUE 
+#define STM32_PLLMUL_VALUE                  4
+#undef STM32_USBSW  
+#define STM32_USBSW                         STM32_USBSW_PCLK
+#undef STM32_ST_USE_TIMER   
+#define STM32_ST_USE_TIMER                  3
```
Comment (zvecr) [inline keyboards/pulsar/xboard_qs/mcuconf.h:39]: You should not need to change this (plus setting it to a time allocated to GPT is wrong)
```suggestion
```

File: keyboards/pulsar/xboard_qs/chconf.h
Comment (zvecr) [reply keyboards/pulsar/xboard_qs/chconf.h:1]: <https://github.com/qmk/qmk_firmware/pull/22984#discussion_r1473753176>

File: keyboards/pulsar/xboard_qs/rules.mk (hunk 4 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/rules.mk:19]: thank you! 
I'll delete it

File: keyboards/pulsar/xboard_qs/mcuconf.h (hunk 7 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/mcuconf.h]: Thank you for your kind consideration

File: keyboards/pulsar/xboard_qs/config.h (hunk 6 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/config.h:21]: I don't know "git" well, and I don't know "QMK" well, so thank you for reviewing it
I checked as much as I could and completed the correction
I'm sure you're busy, but please review it again when the conditions are available

File: keyboards/pulsar/xboard_qs/mcuconf.h (hunk 7 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/mcuconf.h]: as you say
1. Delete "chconf.h" and,
2. GPT is TIM3,
3. "stm32 timer(systick?)" is set to TIM2

Please check.
Thank you very much for your hard work

File: keyboards/pulsar/xboard_qs/xboard_qs.c [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        1, 1, 1,
+        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+        1, 1, 1,   1,   4, 4, 4, 4,   1, 1, 1, 1,    4, 4, 4, 4,    1, 
+        4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1,    1, 1, 1,
+        1, 1, 1,    4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 1,
+        1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,   1,             
+        1,    1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,    1,   
+        1, 1, 1,     4,4,4,     1, 1, 1, 1,    1, 1, 1
+    }
+};
```
Comment (drashna) [inline keyboards/pulsar/xboard_qs/xboard_qs.c:50]: This should be moved to info.json.  `qmk info -f json` can help generate that.

File: keyboards/pulsar/xboard_qs/mcuconf.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+#pragma once
+
+#include_next <mcuconf.h>
+
+#undef STM32_GPT_USE_TIM3
+#define STM32_GPT_USE_TIM3                  TRUE
+
+/* enable SPI2, used for WS2812underglow(RGB BackLight)/WS2812 RGB-Matrix SPI driver, PIN B15 */
```
Comment (drashna) [inline keyboards/pulsar/xboard_qs/mcuconf.h:24]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/pulsar/xboard_qs/info.json [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    "keyboard_name": "Pulsar XBOARD QS",
+    "manufacturer": "Pulsar",
+    "url": "https://pulsargg.kr/",
+    "maintainer": "James Yun",
+
+    "processor": "STM32F072",
+    "bootloader": "stm32-dfu",
+
+    "usb": {
+        "vid": "0x3710",
```
[the remark this answers is not in the document]
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/info.json:11]: 0x3710 is our company number
I will change the "manufacturer" to company name (Aplusx) (not brand name (Pulsar) )
Thank you very much for your detailed consideration

File: keyboards/pulsar/xboard_qs/keymaps/default/keymap.c
[the remark this answers is not in the document]
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/keymaps/default/keymap.c:1]: Oh, I see
I'll try it, but I need time.
Thank you.

File: keyboards/pulsar/xboard_qs/xboard_qs.c (hunk 8 above)
Comment (skylupin03) [reply keyboards/pulsar/xboard_qs/xboard_qs.c:50]: Thank you.
I learn a lot from you.
I'll try

**Pull request closed without merging**